File: generate.py
from argparse import ArgumentParser
import re
from datetime import date

p = ArgumentParser(description='Generate a jeykll blog post template')
p.add_argument('title', help='title of the post, generates slug')
p.add_argument('category', help='category (e.g blog, writeup)')
args = p.parse_args()

template = '''---
layout: %s
title: %s
categories:
- %s
---

Go-go-gadget blog post
'''

def slugify(value):
    "slugify non-crazy strings"
    value = unicode(re.sub('[^\w\s-]', '', value).strip().lower())
    return re.sub('[-\s]+', '-', value)

fn = str(date.today()) + '-' + slugify(args.title) + '.md'
layout = "writeup" if args.category == "writeup" else "post"
post = template % (layout, args.title, args.category)

with open('_posts/%s' % fn, 'w') as fp:
    fp.write(post)
print "generated %s" % fn

File: generate.rb
require 'date'
require 'optparse'

options = {}
opts = OptionParser.new do |opts|
  opts.banner = "Usage:\n\tgenerate.rb title category " + 
    "\n\nOptions:\n\tcategory\tblog, writeup"
end
opts.parse!

if ARGV.empty?
  puts opts.banner
  exit
end

options[:title] = ARGV.shift
options[:category] = ARGV.shift || "blog"
options[:layout] = options[:category] == "writeup" ? "writeup" : "post"

def slug(title)
  title.downcase.gsub(/[^\w]/, " ").strip.gsub(/\s+/, '-')
end

template = '''---
layout: %{layout}
title: "%{title}"
categories:
- %{category}
---

Go-go-gadget blog post
'''

content = template % options
filename = "_posts/#{Date.today}-#{slug(options[:title])}.md"

File.open(filename, "w") do |file|
  file.write(content)
end

system "subl -n . #{filename}"
puts "Crush it!"
